### PR TITLE
T442 Validator as a client of State

### PR DESCRIPTION
**Pull request by ismagin**

### CODE REVIEW

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala [hunk 1]
```diff
@@ -1,37 +1,223 @@
...
+  with LeaseExtendedStateStorageI with AliasExtendedStorageI,
                   settings: ChainParameters) extends State with ScorexLogging {
 
-  override def included(id: Array[Byte]): Option[Int] = storage.included(id, None)
+  def applyAssetIssueReissueBurnTransaction(height: Int)(tx: Transaction): Unit = tx match {
+    case tx: AssetIssuance =>
+      addAsset(tx.assetId, height, tx.id, tx.quantity, tx.reissuable)
+    case tx: BurnTransaction =>
+      burnAsset(tx.assetId, height, tx.id, -tx.amount)
+    case _ =>
```
Comment (phearnot) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: Rewriting all processors as partial functions will help eliminating these match-alls.

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala [hunk 2]
```diff
@@ -1,37 +1,223 @@
...
+
+    val transactionsToRemove: Seq[String] = heightsToRemove.foldLeft(Seq.empty[String]) { (result, h) =>
+      result ++ storage.getTransactions(s"$asset@$h")
+    }
+
+    val keysToRemove = transactionsToRemove.map(t => s"$asset@$t")
+
+    keysToRemove.foreach { key =>
+      storage.removeKey(key)
+    }
```
Comment (phearnot) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]:
```scala
keysToRemove.foreach(storage.remove)
```
Also could be joined with the previous line for brevity.

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala [hunk 3]
```diff
@@ -1,37 +1,223 @@
...
+  }
+
+  def isReissuable(assetId: AssetId): Boolean = {
+    val asset = Base58.encode(assetId)
+    val heights = storage.getHeights(asset)
+
+    heights.lastOption match {
```
Comment (phearnot) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: Replace pattern match with `exists`.

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala [hunk 4]
```diff
@@ -1,37 +1,223 @@
...
+        } else false
+      case None => false
+    }
+  }
+
+  def getAssetName(assetId: AssetId): String = {
+    storage.getTransaction(assetId).flatMap {
+      case tx: IssueTransaction => Some(tx.asInstanceOf[IssueTransaction])
+      case _ => None
+    }.map(tx => new String(tx.name, Charsets.UTF_8)).getOrElse("Unknown")
```
Comment (phearnot) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]:
```scala
storage.getTransaction(assetId).collect {
  case tx: IssueTransaction => new String(tx.name, Charsets.UTF_8)
}.getOrElse("Unknown")
```

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala [hunk 5]
```diff
@@ -1,37 +1,223 @@
...
+            storage.putOrderMatchTxByDay(orderDay, orderIdStr, prev :+ omIdStr)
+          }
+        }
+      }
+
+      def removeObsoleteDays(timestamp: Long): Unit = {
+        val ts = calcStartDay(timestamp)
+        val daysToRemove: List[Long] = storage.savedDaysKeys.filter(t => t < ts)
+        if (daysToRemove.nonEmpty) {
+          synchronized {
```
Comment (phearnot) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: Locking on `this` is generally a bad idea, just saying.

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala [hunk 6]
```diff
@@ -1,37 +1,223 @@
...
+
+    case _ =>
+  }
+
+  def findPrevOrderMatchTxs(order: Order): Set[ExchangeTransaction] = {
+
+    def parseTxSeq(a: Array[String]): Set[ExchangeTransaction] = {
+      a.toSet.flatMap { s: String => Base58.decode(s).toOption }.flatMap { id =>
+        storage.getTransactionBytes(id).flatMap(b => ExchangeTransaction.parseBytes(b).toOption)
+      }
```
Comment (phearnot) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: This could easily be replaced with for-comprehension, which is much more readable.

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala [hunk 7]
```diff
@@ -1,37 +1,223 @@
...
+  }
+
+  def lastAccountPaymentTransaction(account: Account): Option[PaymentTransaction] = {
+    def loop(h: Int, address: AddressString): Option[PaymentTransaction] = {
+      storage.getAccountChanges(address, h) match {
+        case Some(row) =>
+          val accountTransactions = row.reason.flatMap(id => storage.getTransaction(id))
+            .filter(_.isInstanceOf[PaymentTransaction])
+            .map(_.asInstanceOf[PaymentTransaction])
+            .filter(_.sender.address == address)
```
Comment (phearnot) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: Replace with `collect`, see [here](https://github.com/wavesplatform/Waves/pull/182/files#r105203040).

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala [hunk 8]
```diff
@@ -359,56 +462,40 @@ class StoredState(protected[blockchain] val storage: StateStorageI with OrderMat
...
-      case _ =>
-        0L
+          loop(h)
+        case _ =>
+          0L
+      }
     }
+
+    val ls = storage.lastStatesKeys.filter(a => a.length == 35).map(add => add -> balanceAtHeight(add, heightOpt))
+      .filter(b => b._2 != 0).sortBy(_._1).map(b => b._1 -> JsNumber(b._2))
```
Comment (phearnot) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]:
```scala
storage.lastStatesKeys.collect {
  case a if a.length == 35 => a -> balanceAtHeight(a, heightOpt)
} collect {
  case (a, b) if b != 0 => a -> JsNumber(b)
}.sorted
```
(tuples are sorted lexicographically)

File: src/main/scala/scorex/transaction/BalanceChangeCalculator.scala [hunk 9]
```diff
@@ -64,14 +64,14 @@ object BalanceChangeCalculator {
...
-  def effectiveBalanceChanges(state: State)(storage: LeaseExtendedStateStorageI)(tx: Transaction)
+  def effectiveBalanceChanges(state: State)(tx: Transaction)
   : Either[ValidationError, Seq[EffectiveBalanceChange]] = tx match {
     case tx: LeaseTransaction => resolveAddressOrAlias(state)(tx, tx.recipient).map(recipient => {
       Seq(EffectiveBalanceChange(tx.sender, -tx.amount - tx.fee),
         EffectiveBalanceChange(recipient, tx.amount))
     })
     case tx: LeaseCancelTransaction =>
-      val leaseTx = storage.getExistedLeaseTx(tx.leaseId)
+      val leaseTx = state.findTransaction[LeaseTransaction](tx.leaseId).get
```
Comment (ismagin) [inline src/main/scala/scorex/transaction/BalanceChangeCalculator.scala]: The bahavior hasn't changed: previous method threw an exc in this case

File: src/main/scala/scorex/transaction/State.scala [hunk 10]
```diff
@@ -66,44 +68,29 @@ trait State {
...
-  def incrementingTimestampValidator: IncrementingTimestampValidator
+  def effectiveBalanceWithConfirmations(account: Account, confirmations: Int): Long
 
-  def leaseExtendedState: LeaseExtendedState
+  def addAsset(assetId: AssetId, height: Int, transactionId: Array[Byte], quantity: Long, reissuable: Boolean): Unit
 
-  def effectiveBalanceWithConfirmations(account: Account, confirmations: Int): Long
+  def burnAsset(assetId: AssetId, height: Int, transactionId: Array[Byte], quantity: Long): Unit
 
+  def assetRollbackTo(assetId: AssetId, height: Int): Unit
```
Comment (ismagin) [inline src/main/scala/scorex/transaction/State.scala]: I did no renames, I just exposed the method used semi-publicly

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala (hunk 2 above)
Comment (ismagin) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: In this refactoring I tried to change as little logic as possible

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala (hunk 3 above)
Comment (ismagin) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: This is not a new code. A lot of methods like `getAssetQuantity`, `isReissuable`, `getAssetName` etc will hopefully be merged and completely rewritten to  a single `getAssetInfo`, hence minimal changes now

File: src/test/scala/scorex/lagonaki/integration/AliasTransactionSpecification.scala [hunk 11]
```diff
@@ -31,18 +34,28 @@ class AliasTransactionSpecification extends FunSuite with Matchers with Transact
   }
 
   def ensureSenderHasBalance(sender: PrivateKeyAccount): Unit = {
-    val transfer = TransferTransaction.create(None, aliasCreator, sender, 1000000L, 1L, None, 100000L, Array()).right.get
+    val transfer = TransferTransaction.create(None, aliasCreator, sender, 1000000L, TIME, None, 100000L, Array()).right.get
     state.processBlock(TestBlock(Seq(transfer))).get
     ()
   }
 
+  def shouldBeValid(t: Transaction): Assertion = {
```
Comment (ismagin) [inline src/test/scala/scorex/lagonaki/integration/AliasTransactionSpecification.scala]: agreed it's not worth reusing, time and validators are different

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala (hunk 1 above)
Comment (ismagin) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: Won't work in general since they intersect

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala (hunk 2 above)
Comment (ismagin) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: Fixed

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala (hunk 4 above)
Comment (ismagin) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: reworked to extension method over `findTransaction`

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala (hunk 6 above)
Comment (ismagin) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: fixed

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala (hunk 8 above)
Comment (ismagin) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: This doesn't compile, `lastStatesKeys` is a `List[String]`?

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala (hunk 7 above)
Comment (ismagin) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: replaced with `findTransaction`

File: src/main/scala/scorex/transaction/BalanceChangeCalculator.scala [hunk 12]
```diff
@@ -64,14 +64,14 @@ object BalanceChangeCalculator {
     }
   }
 
-  def effectiveBalanceChanges(state: State)(storage: LeaseExtendedStateStorageI)(tx: Transaction)
+  def effectiveBalanceChanges(state: State)(tx: Transaction)
   : Either[ValidationError, Seq[EffectiveBalanceChange]] = tx match {
```
Comment (ismagin) [inline src/main/scala/scorex/transaction/BalanceChangeCalculator.scala]: fixed

File: src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala (hunk 8 above)
Comment (ismagin) [inline src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala]: The requested changes are out of scope
